pkg/armhelpers/deploymentError.go

```diff
@@ -0,0 +1,150 @@
+		// chances of multiple deployment errors being returned for a single resource type is slim
+		// (but possible) based on current error/QoS analysis. In those cases where multiple errors
+		// are returned ACS will pick the first error code for determining whether this is an internal
+		// or a client error. This can be reevaluated later based on practical experience.
+		// However, note that customer will be returned the entire contents of "StatusMessage" object
+		// (like before) so they have access to all the errors returned by ARM.
+		logger.Infof("Found ResourceDeploymentFailure error code - error response = '%v'", *errresp)
+		details := errresp.Body.Details
+		if len(details) > 0 {
+			errresp.Body.Code = details[0].Code
+			errresp.Body.Message = details[0].Message
+			errresp.Body.Target = details[0].Target
+		}
+	}
+
+	return errresp, nil
+}
+
+func toArmErrors(logger *logrus.Entry, deploymentName string, operationsList resources.DeploymentOperationsListResult) ([]*ErrorResponse, error) {
+	ret := []*ErrorResponse{}
+
+	if operationsList.Value == nil {
+		return ret, nil
+	}
+
+	for _, operation := range *operationsList.Value {
+		if operation.Properties == nil || *operation.Properties.ProvisioningState != string(api.Failed) {
+			continue
+		}
+
+		errresp, err := toArmError(logger, operation)
+		if err != nil {
+			logger.Errorf("unable to convert deployment operation to error response in deployment %s from ARM. error: %v", deploymentName, err)
+			return ret, err
+		}
+
+		if len(errresp.Body.Code) > 0 {
+			logger.Warnf("got failed deployment operation in deployment %s. error: %v", deploymentName, errresp.Error())
+		}
+		ret = append(ret, errresp)
+	}
+	return ret, nil
+}
+
+// GetDeploymentError returns deployment error
+func GetDeploymentError(az ACSEngineClient, logger *logrus.Entry, resourceGroupName, deploymentName string) (*ErrorResponse, error) {
+	errList := []*ErrorResponse{}
+	logger.Infof("Getting detailed deployment errors for %s", deploymentName)
+
+	var top int32 = 1
+	operationList, err := az.ListDeploymentOperations(resourceGroupName, deploymentName, &top)
+	if err != nil {
+		logger.Warnf("unable to list deployment operations: %v", err)
+		return nil, err
+	}
+	eList, err := toArmErrors(logger, deploymentName, operationList)
+	if err != nil {
+		return nil, err
+	}
+	errList = append(errList, eList...)
+	for operationList.NextLink != nil {
+		operationList, err = az.ListDeploymentOperationsNextResults(operationList)
+		if err != nil {
+			logger.Warnf("unable to list next deployment operations: %v", err)
+			break
+		}
+		eList, err := toArmErrors(logger, deploymentName, operationList)
+		if err != nil {
+			return nil, err
+		}
+		errList = append(errList, eList...)
+	}
+	switch len(errList) {
+	case 0:
+		return nil, fmt.Errorf("unable to extract deployment error for %s", deploymentName)
+	case 1:
+		return errList[0], nil
+	default:
+		// combine all errors
+		for _, e := range errList[1:] {
```

Comment:
why cant we just return an array of `*ErrorResponse`? Why do we want to change it?